### DISCUSSION TITLE
Is it possible to fetch only on server-side?

### DISCUSSION THREAD

**Discussion by TotomInc** [nuxt/nuxt#17291] [Questions]
Hi, I would like to know if it's possible to fetch only on server-side with Nuxt 3 (RC 1).

I actually want to fetch my content from an headless CMS only during server-side rendering and never do API calls to my CMS on client-side (because it's slow and resource-intensive).

Here's what I tried:

```vue
// pages/index.vue
<script setup lang="ts">
import type { HomepageResponse } from "~/interfaces/homepage.interfaces";

const homepage = process.server
  ? await $fetch<HomepageResponse>("http://localhost:1337/api/homepage?populate=deep")
    .then((res) => res.data.attributes)
  : null;

// homepage will have the proper content but after a few ms it takes the `null` value.
console.log(homepage);
</script>
```

```
runtime-core.esm-bundler.js:38 [Vue warn]: Hydration node mismatch:
- Client vnode: Symbol(Comment) 
- Server rendered DOM: <!--[--> (start of fragment) 
  at <Index onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > key="/" > 
  at <BaseTransition mode="out-in" appear=false persisted=false  ... > 
  at <Transition name="page" mode="out-in" > 
  at <RouterView > 
  at <NuxtPage> 
  at <DefaultLayout > 
  at <AsyncComponentWrapper > 
  at <BaseTransition mode="out-in" appear=false persisted=false  ... > 
  at <Transition name="layout" mode="out-in" > 
  at <Anonymous name="default-layout" > 
  at <App key=1 > 
  at <NuxtRoot>
```

```typescript
// nuxt.config.ts
import { defineNuxtConfig } from 'nuxt';

// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
  target: process.env.NODE_ENV === 'production' ? 'static' : 'server',
});
```

The thing I need is to make the request server-side, blocking the page loading, until the server has successfully fetched the response and renderer only in the server.

I read v3 documentation about data-fetching, about the $fetch API but also the v2 blog-post about going [_full-static_](https://nuxtjs.org/announcements/going-full-static/) (which is what I'm trying to achieve) but without success.

**Comment by TotomInc** [comment]
It looks like the issue is coming from using vue-router with `<NuxtLink />`.

If I use `<a href>` instead of `<NuxtLink />`, the page will instead use the payload that has been generated, and not ask for an API call.

By using an `<a>`, it means we are not using an SPA anymore, which cause re-loading all assets (fonts, vendors, etc.) when navigating. This is not really performant, especially if you have a lot of assets to load.

This fix is highly inspired from [this StackOverflow answer](https://stackoverflow.com/a/61355674/3669565).

Here is my actual setup:

```typescript
// nuxt.config.ts
import { defineNuxtConfig } from "nuxt";

// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
  target: "static",
});
```

```typescript
// pages/product/[slug].vue
const { data } = await useAsyncData("product", () =>
  $fetch<StrapiGetProducts>(
    `http://localhost:1337/api/products?filters[slug][$eq]=${route.params.slug}&populate=deep`
  )
);

// Use <a /> to do navigation on your Vue template, instead of <NuxtLink />.
```

I would really like to find a fix to make my SPA loading only generated the payload, without requesting my back-end when navigating with vue-router (`<NuxtLink />`).

However, what I still don't understand, as stated in the ["Going Full Static" announcement for Nuxt v2](https://nuxtjs.org/announcements/going-full-static/#crazy-fast-static-applications), is:

> `nuxt generate` with `target: 'static'` will pre-render all your pages to HTML and save a payload file in order to mock `asyncData` and `fetch` on client-side navigation, this means no more HTTP calls to your API on client-side navigation. By extracting the page payload to a js file, it also reduces the HTML size served as well as preloading it (from the in the header) for optimal performance.

When using `useAsyncData` with `$fetch` and `target: static`, navigating with `<NuxtLink />`, I still have HTTP calls to my API during client-side navigation, which should not be the expected behaviour.

The payload file is properly generated too, there are no issues with it. But it seems like Nuxt prefer to create an HTTP call instead of forcing the usage of the payload file when navigating at least once in the app.

Maybe this is still not implemented in Nuxt v3?

**Reply by TotomInc** [reply]
Sorry for bothering you, @Atinux and @danielroe, but I don't know if this should still be a discussion or can be turned into an issue. 🙏 

I can provide my repository source-code, if needed to help understanding.

**Reply by danielroe** [reply]
At the moment, if you wish this behaviour, move your code into a Nitro API route: `~/server/api/products.ts`. Within that route, you can fetch from your backend/cache. On frontend, you can request via `$fetch('/api/products')` and it will not expose your token.

A full static mode is still coming that will stub out those `useAsyncData` requests in future.

**Reply by TotomInc** [reply]
Thanks for your answer.

How should I do the fetching on the `~/server/api/products.ts`?

I tried by moving the `useAsyncData` with `$fetch` inside `server/api/products.ts` and calling `$fetch('/api/products')` on my `pages/products.vue`:

```typescript
// server/api/products.ts
import type { StrapiGetProducts } from "@/interfaces/strapi.interfaces";

export default defineEventHandler(async () => {
  const { data } = await useAsyncData("product", () =>
    $fetch<StrapiGetProducts>(
      `http://localhost:1337/api/products?filters[slug][$eq]=${route.params.slug}&populate=deep`
    )
  );

  return data;
});
```

But it looks like I am unable to import `useAsyncData` in server-routes, it causes a 500.

When I try without `useAsyncData` but only `$fetch`, the products payload doesn't seem to be generated and saved to any file:

```typescript
// server/api/products.ts
import type { StrapiGetProducts } from "@/interfaces/strapi.interfaces";

export default defineEventHandler(async () => {
  const data = await $fetch<StrapiGetProducts>(
    `http://localhost:1337/api/products?populate=deep`
  );

  return data;
});

```

The client-side still tries to do an HTTP call to `/api/products` this time.

I think there is a lot of confusion right now with fully static website generation and data fetching with the Nuxt docs.

I understand the v3 is still in release candidate, that's why I am also trying to do fully static in order to make things clear and potentially contribute to the documentation later. 🙂

**Reply by TotomInc** [reply]
After trying for a few hours the last days, it looks like it doesn't really work well.

I haven't found an alternative solution to make the website entirely static, without relying on the back-end during client-side navigation: the client-side always want to do re-hydration when changing page with vue-router, it doesn't want to use the generated payload, but forces me to do an API call.

Another pain point: the plugin API doesn't seems to be stable too (I tried to use `generate:distCopied` from v2, but it looks like there are no alternative hook, for example).

Nuxt v3 doesn't seem to be built around the SSG fully-static concept, so I may give it a few months again before I can pick it up again.

Thanks for the hard work of the team, hope v3 will be released in stable soon! 🤞

**Reply by uuf6429** [reply]
Hey, is this still the case?
To be clear, in my case, the site is properly rendered, but it still calls the api on the frontend (even though the html is fully rendered and there's a copy of the api payload data). When I block the api request, it's ends up as a blank page. Interestingly, when I run this in dev mode (`nuxi dev`), there are no api calls on the frontend.

**Reply by Patrick-Remy** [reply]
> in my case, the site is properly rendered, but it still calls the api on the frontend (even though the html is fully rendered and there's a copy of the api payload data). When I block the api request, it's ends up as a blank page

I can confirm the same behavior.

**Reply by danielroe** [reply]
Would you share some code or raise an issue with a reproduction? It should definitely not do that.

**Reply by TotomInc** [reply]
@uuf6429 and @Patrick-Remy, did your fetch implementation looks like this?

Because that's what I actually have, and is not working properly.

```typescript
// pages/product/[slug].vue
// Should be fetched only during rendering, when generating website.
// useAsyncData should be smart enough to cache the data in the Nuxt payload (I think)
// and retrieve it during client-side navigation (spoiler: it doesn't 😄)
const { data } = await useAsyncData("product", () =>
  $fetch<StrapiGetProducts>(
    `http://localhost:1337/api/products?filters[slug][$eq]=${route.params.slug}&populate=deep`
  )
);

// Consume the fetch response from useAsyncData.
const product = data.value.data[0].attributes;
```

If I go to productionwebsite.com/product/my-product, fetch is not called and payload from Nuxt is retrieved.

But the issue occurs when I am doing at least 1 client-side navigation. If I go now to productionwebsite.com/product/my-other-product, then a fetch will be called instead of retrieving the Nuxt payload.

**Reply by danielroe** [reply]
Yes, stubbing `useAsyncData` responses is not yet implemented. That's what a 'full static' eqiuvalent might be in Nuxt 3. (Though this is distinct from SSG, which is already fully in place.)

**Reply by uuf6429** [reply]
@TotomInc @danielroe Thanks for the clarification...if I might add something, (that still confuses me a bit), what is the usefulness of the current approach?

In the sense that if static generation is disabled, with SSR the data would be retrieved server-side, returned to the client side, and then re-retrieved on the client side again. Or is this flow only happening for static generation?

Additionally, I find this a bit weird (conceptually) that it works for useFetch but not useAsyncData...could it be that it was implemented in the wrong place?

**Reply by danielroe** [reply]
@uuf6429 No, with the current approach, it does not re-fetch on client-side load. Are you setting your key correctly in `useAsyncData`?

**Reply by uuf6429** [reply]
> Are you setting your key correctly in useAsyncData?

@danielroe yes, I haven't tried the SSR approach (I think), I was only assuming the same logic would follow.

**Comment by hagemann** [comment]
Using plain vanilla [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) like this should work (I had to solve something similar):

```
// server/api/products.ts
export default defineEventHandler(async () => {
  var result = await fetch(`http://localhost:1337/api/products?populate=deep`);
  var data = await result.json()

  return data;
});
```
